python/ray/monitor.py

```diff
@@ -468,6 +475,77 @@ def driver_removed_handler(self, unused_channel, data):
 
         self._clean_up_entries_for_driver(driver_id)
 
+    def _xray_clean_up_entries_for_driver(self, driver_id):
+        """Remove this driver's object/task entries from redis.
+
+        Removes control-state entries of all tasks and task return
+        objects belonging to the driver.
+        """
+        redis = self.state.redis_client
+
+        task_table_infos = {}  # task id -> TaskInfo
+        for key in redis.scan_iter(match=XRAY_TASK_TABLE_PREFIX + b"*"):
+            entry = redis.get(key)
+            task = Task.GetRootAsTask(entry, 0)
+            task_info = TaskInfo.GetRootAsTaskInfo(task.TaskSpecification(), 0)
+            if driver_id != task_info.DriverId():
+                # Ignore tasks that aren't from this driver.
+                continue
+            task_table_infos[task_info.TaskId()] = task_info
+        driver_task_ids = task_table_infos.keys()
+
+        # Get the list of objects returned by driver tasks.
+        driver_object_ids = []
+        for task_info in task_table_infos.values():
+            driver_object_ids.extend([
+                task_info.Returns(i) for i in range(task_info.ReturnsLength())
+            ])
+
+        # TODO: Remove only objects in GCS that originated from this driver.
+        # In legacy Ray, put objects are added by local_scheduler.cc
+        # whenever a message of type ray::local_scheduler::protocol::PutObject
+        # is received. In XRay, object locations are registered with the GCS
+        # whenever the ObjectManager receives notifications about local objects
+        # from the object store. Enable this for XRay by implementing the PutObject
+        # protocol in NodeManager, and a separate object table for tracking put
+        # information.
+
+        # Clean up entries for non-empty objects.
+        non_empty_objects = []
+        for object_id in driver_object_ids:
+            entry = redis.zrange(XRAY_OBJECT_TABLE_PREFIX + object_id, 0, 0)
+            if entry:
+                non_empty_objects.append(object_id)
+
+        # Form the redis keys to delete.
+        keys = [XRAY_TASK_TABLE_PREFIX + k for k in driver_task_ids]
+        keys.extend([XRAY_OBJECT_TABLE_PREFIX + k for k in non_empty_objects])
+
+        if not keys:
+            return
+
+        # Remove with best effort.
+        num_deleted = redis.delete(*keys)
+        log.info(
+            "Removed {} dead redis entries of the driver from redis.".
+                format(num_deleted))
+        if num_deleted != len(keys):
+            log.warning(
+                "Failed to remove {} relevant "
+                "redis entries.".format(len(keys) - num_deleted))
+
+    def xray_driver_removed_handler(self, unused_channel, data):
+        """Handle a notification that a driver has been removed.
+        """
+        # TODO: Migrate this to raylet monitor.
```

Comment:
Put your name in the TODO so that people know who wrote it.

However, do we actually want this in the raylet monitor? The raylet monitor is intended for detecting liveness of machines, whereas this functionality is more about resource management and cleanup.